File: src/main/java/com/github/simiacryptus/aicoder/text/BlockComment.java
package com.github.simiacryptus.aicoder.text;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;

public class BlockComment extends IndentedText {
    public static class Factory implements TextBlockFactory<BlockComment> {
        public final String linePrefix;
        public final String blockPrefix;
        public final String blockSuffix;

        public Factory(String blockPrefix, String linePrefix, String blockSuffix) {
            this.linePrefix = linePrefix;
            this.blockPrefix = blockPrefix;
            this.blockSuffix = blockSuffix;
        }

        @Override
        public BlockComment fromString(String text) {
            text = StringTools.stripSuffix(StringTools.trimSuffix(text.replace("\t", TAB_REPLACEMENT)), blockSuffix.trim());
            String indent = StringTools.getWhitespacePrefix(text.split(DELIMITER));
            return new BlockComment(blockPrefix, linePrefix, blockSuffix, indent, Arrays.stream(text.split(DELIMITER))
                    .map(s -> StringTools.stripPrefix(s, indent))
                    .map(StringTools::trimPrefix)
                    .map(s -> StringTools.stripPrefix(s, blockPrefix.trim()))
                    .map(s -> StringTools.stripPrefix(s, linePrefix.trim()))
                    .toArray(String[]::new));
        }

        @Override
        public boolean looksLike(String text) {
            return text.trim().startsWith(blockPrefix) && text.trim().endsWith(blockSuffix);
        }
    }

    public final String linePrefix;
    public final String blockPrefix;
    public final String blockSuffix;


    public BlockComment(String blockPrefix, String linePrefix, String blockSuffix, String indent, String... textBlock) {
        super(indent, textBlock);
        this.linePrefix = linePrefix;
        this.blockPrefix = blockPrefix;
        this.blockSuffix = blockSuffix;
    }

    @Override
    public @NotNull String toString() {
        String indent = getIndent();
        String delimiter = DELIMITER + indent;
        String joined = Arrays.stream(rawString()).map(x->linePrefix + " " + x).collect(Collectors.joining(delimiter));
        return blockPrefix + delimiter + joined + delimiter + blockSuffix;
    }

    public @NotNull IndentedText withIndent(String indent) {
        return new BlockComment(blockPrefix, linePrefix, blockSuffix, indent, textBlock);
    }
}

File: src/main/kotlin/com/simiacryptus/aicoder/util/BlockComment.kt
@file:Suppress("NAME_SHADOWING")

package com.simiacryptus.aicoder.util

import com.simiacryptus.aicoder.util.TextBlock.Companion.DELIMITER
import com.simiacryptus.aicoder.util.TextBlock.Companion.TAB_REPLACEMENT
import com.simiacryptus.util.StringUtil
import java.util.*
import java.util.stream.Collectors

class BlockComment(
  private val blockPrefix: CharSequence,
  private val linePrefix: CharSequence,
  private val blockSuffix: CharSequence,
  indent: CharSequence,
  vararg textBlock: CharSequence
) :
  IndentedText(indent, *textBlock) {
  class Factory(private val blockPrefix: String, private val linePrefix: String, private val blockSuffix: String) :
    TextBlockFactory<BlockComment?> {
    override fun fromString(text: String?): BlockComment {
      var text = text!!
      text = StringUtil.stripSuffix(
        StringUtil.trimSuffix(text.replace("\t", TAB_REPLACEMENT.toString(), false)),
        blockSuffix.trim { it <= ' ' })
      val indent = StringUtil.getWhitespacePrefix(*text.split(DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }
        .toTypedArray())
      return BlockComment(
        blockPrefix,
        linePrefix,
        blockSuffix,
        indent,
        *(Arrays.stream(text.split(DELIMITER.toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray())
          .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, indent) }
          .map { text: CharSequence? -> StringUtil.trimPrefix(text!!) }
          .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, blockPrefix.trim { it <= ' ' }) }
          .map { s: CharSequence? -> StringUtil.stripPrefix(s!!, linePrefix.trim { it <= ' ' }) }
          .collect(Collectors.toList()).toTypedArray()))
    }

    override fun looksLike(text: String?): Boolean {
      return text!!.trim { it <= ' ' }.startsWith(blockPrefix) && text.trim { it <= ' ' }.endsWith(blockSuffix)
    }
  }

  override fun toString(): String {
    val indent = this.indent
    val delimiter: CharSequence = DELIMITER + indent
    val joined: CharSequence = Arrays.stream(rawString()).map { x: CharSequence -> "$linePrefix $x" }
      .collect(Collectors.joining(delimiter))
    return blockPrefix.toString() + delimiter + joined + delimiter + blockSuffix
  }

  override fun withIndent(indent: CharSequence): IndentedText {
    return BlockComment(blockPrefix, linePrefix, blockSuffix, indent, *lines)
  }
}
